File: server/src/main/java/de/ingrid/igeserver/exports/ExporterFactory.java
package de.ingrid.igeserver.exports;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class ExporterFactory {

    final List<IgeExporter> exporterList;

    @Autowired
    public ExporterFactory(List<IgeExporter> exporterList) {
        this.exporterList = exporterList;
    }

    public List<ExportTypeInfo> getTypeInfos() {
		return exporterList.stream()
				.map(IgeExporter::getTypeInfo)
				.collect(Collectors.toList());
	}

    public IgeExporter getExporter(String format) {
        return this.exporterList.stream()
				.filter(e -> format.equals(e.getTypeInfo().getType()))
				.findAny()
				.orElse(null);
    }
}

File: server/src/main/java/de/ingrid/igeserver/exports/ExporterFactory.kt
package de.ingrid.igeserver.exports

import de.ingrid.igeserver.configuration.ConfigurationException
import de.ingrid.igeserver.services.DocumentCategory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class ExporterFactory(private val exporterList: List<IgeExporter>) {
    val typeInfos: List<ExportTypeInfo> = exporterList.map { exporter: IgeExporter -> exporter.typeInfo }

    fun getExporter(category: DocumentCategory, format: String): IgeExporter {
        try {
            return exporterList.first { exporter: IgeExporter -> category == exporter.typeInfo.category && format == exporter.typeInfo.type }
        } catch (e: NoSuchElementException) {
            throw ConfigurationException.withReason("No exporter found for format '$format'.")
        }
    }
}